File: lwjgl/lesson20/Lesson20/Lesson20.java
/*
 *      This Code Was Created By Jeff Molofee and GB Schmick 2000
 *      A HUGE Thanks To Fredric Echols For Cleaning Up
 *      And Optimizing The Base Code, Making It More Flexible!
 *      If You've Found This Code Useful, Please Let Me Know.
 *      Visit Our Sites At www.tiptup.com and nehe.gamedev.net
 */

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.IntBuffer;

import org.lwjgl.opengl.Display;
import org.lwjgl.opengl.DisplayMode;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.glu.GLU;
import org.lwjgl.devil.IL;
import org.lwjgl.input.Keyboard;

/**
 * @author Mark Bernard
 * date:    26-Jun-2004
 *
 * Port of NeHe's Lesson 20 to LWJGL
 * Title: Masking
 * Uses version 0.9alpha of LWJGL http://www.lwjgl.org/
 *
 * Be sure that the LWJGL libraries are in your classpath
 *
 * Ported directly from the C++ version
 *
 * 2004-10-08: Updated to version 0.92alpha of LWJGL.
 * 2004-12-19: Updated to version 0.94alpha of LWJGL and to use
 *             DevIL for image loading.
 */
public class Lesson20 {
    private boolean done = false;
    private boolean fullscreen = false;
    private final String windowTitle = "NeHe's OpenGL Lesson 20 for LWJGL (Masking)";
    private boolean f1 = false; // F1 key pressed
    private DisplayMode displayMode;

    boolean masking = true;     // Masking On/Off
    boolean mp;                 // M Pressed?
    boolean sp;                 // Space Pressed?
    boolean scene;              // Which Scene To Draw

    int  texture[];             // Storage For Our Five Textures

    float roll;                 // Rolling Texture


    public static void main(String args[]) {
        boolean fullscreen = false;
        if(args.length>0) {
            if(args[0].equalsIgnoreCase("fullscreen")) {
                fullscreen = true;
            }
        }

        Lesson20 l20 = new Lesson20();
        l20.run(fullscreen);
    }
    public void run(boolean fullscreen) {
        this.fullscreen = fullscreen;
        try {
            init();
            while (!done) {
                mainloop();
                render();
                Display.update();
            }
            cleanup();
        }
        catch (Exception e) {
            e.printStackTrace();
            System.exit(0);
        }
    }
    private void mainloop() {
        if(Keyboard.isKeyDown(Keyboard.KEY_ESCAPE)) {       // Exit if Escape is pressed
            done = true;
        }
        if(Display.isCloseRequested()) {                     // Exit if window is closed
            done = true;
        }
        if(Keyboard.isKeyDown(Keyboard.KEY_F1) && !f1) {    // Is F1 Being Pressed?
            f1 = true;                                      // Tell Program F1 Is Being Held
            switchMode();                                   // Toggle Fullscreen / Windowed Mode
        }
        if(!Keyboard.isKeyDown(Keyboard.KEY_F1)) {          // Is F1 Being Pressed?
            f1 = false;
        }
        if(Keyboard.isKeyDown(Keyboard.KEY_SPACE) && !sp) { // Is Space Being Pressed?
            sp = true;                                      // Tell Program Spacebar Is Being Held
            scene = !scene;                                 // Toggle From One Scene To The Other
        }
        if(!Keyboard.isKeyDown(Keyboard.KEY_SPACE)) {       // Has Spacebar Been Released?
            sp = false;                                     // Tell Program Spacebar Has Been Released
        }
        if(Keyboard.isKeyDown(Keyboard.KEY_M) && !mp) {     // Is M Being Pressed?
            mp = true;                                      // Tell Program M Is Being Held
            masking = !masking;                             // Toggle Masking Mode OFF/ON
        }
        if(!Keyboard.isKeyDown(Keyboard.KEY_M)) {           // Has M Been Released?
            mp = false;                                     // Tell Program That M Has Been Released
        }
    }

    private void switchMode() {
        fullscreen = !fullscreen;
        try {
            Display.setFullscreen(fullscreen);
        }
        catch(Exception e) {
            e.printStackTrace();
        }
    }

    private void createWindow() throws Exception {
        Display.setFullscreen(fullscreen);
        DisplayMode d[] = Display.getAvailableDisplayModes();
        for (int i = 0; i < d.length; i++) {
            if (d[i].getWidth() == 640
                && d[i].getHeight() == 480
                && d[i].getBitsPerPixel() == 32) {
                displayMode = d[i];
                break;
            }
        }
        Display.setDisplayMode(displayMode);
        Display.setTitle(windowTitle);
        Display.create();
    }
    private void init() throws Exception {
        createWindow();
        IL.create();

        loadTextures();
        initGL();
    }
    private void loadTextures() {                                    // Load Bitmaps And Convert To Textures
        texture = new int[5];
        texture[0] = loadTexture("Data/Logo.bmp");
        texture[1] = loadTexture("Data/Mask1.bmp");
        texture[2] = loadTexture("Data/Image1.bmp");
        texture[3] = loadTexture("Data/Mask2.bmp");
        texture[4] = loadTexture("Data/Image2.bmp");
    }
    private void initGL() {                                     // All Setup For OpenGL Goes Here
        GL11.glEnable(GL11.GL_TEXTURE_2D);                            // Enable Texture Mapping
        GL11.glShadeModel(GL11.GL_SMOOTH);                            // Enable Smooth Shading
        GL11.glClearColor(0.0f, 0.0f, 0.0f, 0.5f);               // Black Background
        GL11.glClearDepth(1.0f);                                 // Depth Buffer Setup
        GL11.glEnable(GL11.GL_DEPTH_TEST);                            // Enables Depth Testing
        GL11.glMatrixMode(GL11.GL_PROJECTION); // Select The Projection Matrix
        GL11.glLoadIdentity(); // Reset The Projection Matrix

        // Calculate The Aspect Ratio Of The Window
        GLU.gluPerspective(45.0f,
                (float) displayMode.getWidth() / (float) displayMode.getHeight(),
                0.1f,100.0f);
        GL11.glMatrixMode(GL11.GL_MODELVIEW); // Select The Modelview Matrix
        GL11.glLoadIdentity(); // Reset The Modelview Matrix

        // Really Nice Perspective Calculations
        GL11.glHint(GL11.GL_PERSPECTIVE_CORRECTION_HINT, GL11.GL_NICEST);
    }

    private void render() {
        GL11.glClear(GL11.GL_COLOR_BUFFER_BIT | GL11.GL_DEPTH_BUFFER_BIT);     // Clear Screen And Depth Buffer
        GL11.glLoadIdentity();                                       // Reset The ModelView Matrix
        GL11.glTranslatef(0.0f,0.0f,-2.0f);                      // Move Into The Screen 5 Units

        GL11.glBindTexture(GL11.GL_TEXTURE_2D, texture[0]);           // Select Our Logo Texture
        GL11.glBegin(GL11.GL_QUADS);                                  // Start Drawing A Textured Quad
            GL11.glTexCoord2f(0.0f, -roll+0.0f); GL11.glVertex3f(-1.1f, -1.1f,  0.0f);    // Bottom Left
            GL11.glTexCoord2f(3.0f, -roll+0.0f); GL11.glVertex3f( 1.1f, -1.1f,  0.0f);    // Bottom Right
            GL11.glTexCoord2f(3.0f, -roll+3.0f); GL11.glVertex3f( 1.1f,  1.1f,  0.0f);    // Top Right
            GL11.glTexCoord2f(0.0f, -roll+3.0f); GL11.glVertex3f(-1.1f,  1.1f,  0.0f);    // Top Left
        GL11.glEnd();                                            // Done Drawing The Quad

        GL11.glEnable(GL11.GL_BLEND);                                 // Enable Blending
        GL11.glDisable(GL11.GL_DEPTH_TEST);                           // Disable Depth Testing

        if (masking)                                        // Is Masking Enabled?
        {
            GL11.glBlendFunc(GL11.GL_DST_COLOR,GL11.GL_ZERO);              // Blend Screen Color With Zero (Black)
        }

        if (scene)                                          // Are We Drawing The Second Scene?
        {
            GL11.glTranslatef(0.0f,0.0f,-1.0f);                  // Translate Into The Screen One Unit
            GL11.glRotatef(roll*360,0.0f,0.0f,1.0f);             // Rotate On The Z Axis 360 Degrees.
            if (masking)                                    // Is Masking On?
            {
                GL11.glBindTexture(GL11.GL_TEXTURE_2D, texture[3]);   // Select The Second Mask Texture
                GL11.glBegin(GL11.GL_QUADS);                          // Start Drawing A Textured Quad
                    GL11.glTexCoord2f(0.0f, 0.0f); GL11.glVertex3f(-1.1f, -1.1f,  0.0f);  // Bottom Left
                    GL11.glTexCoord2f(1.0f, 0.0f); GL11.glVertex3f( 1.1f, -1.1f,  0.0f);  // Bottom Right
                    GL11.glTexCoord2f(1.0f, 1.0f); GL11.glVertex3f( 1.1f,  1.1f,  0.0f);  // Top Right
                    GL11.glTexCoord2f(0.0f, 1.0f); GL11.glVertex3f(-1.1f,  1.1f,  0.0f);  // Top Left
                GL11.glEnd();                                    // Done Drawing The Quad
            }

            GL11.glBlendFunc(GL11.GL_ONE, GL11.GL_ONE);                    // Copy Image 2 Color To The Screen
            GL11.glBindTexture(GL11.GL_TEXTURE_2D, texture[4]);       // Select The Second Image Texture
            GL11.glBegin(GL11.GL_QUADS);                              // Start Drawing A Textured Quad
                GL11.glTexCoord2f(0.0f, 0.0f); GL11.glVertex3f(-1.1f, -1.1f,  0.0f);  // Bottom Left
                GL11.glTexCoord2f(1.0f, 0.0f); GL11.glVertex3f( 1.1f, -1.1f,  0.0f);  // Bottom Right
                GL11.glTexCoord2f(1.0f, 1.0f); GL11.glVertex3f( 1.1f,  1.1f,  0.0f);  // Top Right
                GL11.glTexCoord2f(0.0f, 1.0f); GL11.glVertex3f(-1.1f,  1.1f,  0.0f);  // Top Left
            GL11.glEnd();                                        // Done Drawing The Quad
        }
        else                                                // Otherwise
        {
            if (masking)                                    // Is Masking On?
            {
                GL11.glBindTexture(GL11.GL_TEXTURE_2D, texture[1]);   // Select The First Mask Texture
                GL11.glBegin(GL11.GL_QUADS);                          // Start Drawing A Textured Quad
                    GL11.glTexCoord2f(roll+0.0f, 0.0f); GL11.glVertex3f(-1.1f, -1.1f,  0.0f); // Bottom Left
                    GL11.glTexCoord2f(roll+4.0f, 0.0f); GL11.glVertex3f( 1.1f, -1.1f,  0.0f); // Bottom Right
                    GL11.glTexCoord2f(roll+4.0f, 4.0f); GL11.glVertex3f( 1.1f,  1.1f,  0.0f); // Top Right
                    GL11.glTexCoord2f(roll+0.0f, 4.0f); GL11.glVertex3f(-1.1f,  1.1f,  0.0f); // Top Left
                GL11.glEnd();                                    // Done Drawing The Quad
            }

            GL11.glBlendFunc(GL11.GL_ONE, GL11.GL_ONE);                    // Copy Image 1 Color To The Screen
            GL11.glBindTexture(GL11.GL_TEXTURE_2D, texture[2]);       // Select The First Image Texture
            GL11.glBegin(GL11.GL_QUADS);                              // Start Drawing A Textured Quad
                GL11.glTexCoord2f(roll+0.0f, 0.0f); GL11.glVertex3f(-1.1f, -1.1f,  0.0f); // Bottom Left
                GL11.glTexCoord2f(roll+4.0f, 0.0f); GL11.glVertex3f( 1.1f, -1.1f,  0.0f); // Bottom Right
                GL11.glTexCoord2f(roll+4.0f, 4.0f); GL11.glVertex3f( 1.1f,  1.1f,  0.0f); // Top Right
                GL11.glTexCoord2f(roll+0.0f, 4.0f); GL11.glVertex3f(-1.1f,  1.1f,  0.0f); // Top Left
            GL11.glEnd();                                        // Done Drawing The Quad
        }

        GL11.glEnable(GL11.GL_DEPTH_TEST);                            // Enable Depth Testing
        GL11.glDisable(GL11.GL_BLEND);                                // Disable Blending

        roll+=0.002f;                                       // Increase Our Texture Roll Variable
        if (roll>1.0f)                                      // Is Roll Greater Than One
        {
            roll-=1.0f;                                     // Subtract 1 From Roll
        }
    }

    private void cleanup() {
        Display.destroy();
    }

    private int loadTexture(String path) {
        IntBuffer image = ByteBuffer.allocateDirect(4).order(ByteOrder.nativeOrder()).asIntBuffer();
        IL.ilGenImages(1, image);
        IL.ilBindImage(image.get(0));
        IL.ilLoadImage(path);
        IL.ilConvertImage(IL.IL_RGB, IL.IL_BYTE);
        ByteBuffer scratch = ByteBuffer.allocateDirect(IL.ilGetInteger(IL.IL_IMAGE_WIDTH) * IL.ilGetInteger(IL.IL_IMAGE_HEIGHT) * 3);
        IL.ilCopyPixels(0, 0, 0, IL.ilGetInteger(IL.IL_IMAGE_WIDTH), IL.ilGetInteger(IL.IL_IMAGE_HEIGHT), 1, IL.IL_RGB, IL.IL_BYTE, scratch);
        
        // Create A IntBuffer For Image Address In Memory
        IntBuffer buf = ByteBuffer.allocateDirect(4).order(ByteOrder.nativeOrder()).asIntBuffer();
        GL11.glGenTextures(buf); // Create Texture In OpenGL

        GL11.glBindTexture(GL11.GL_TEXTURE_2D, buf.get(0));
        // Typical Texture Generation Using Data From The Image

        // Linear Filtering
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, GL11.GL_LINEAR);
        // Linear Filtering
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MAG_FILTER, GL11.GL_LINEAR);
        // Generate The Texture
        GL11.glTexImage2D(GL11.GL_TEXTURE_2D, 0, GL11.GL_RGB, IL.ilGetInteger(IL.IL_IMAGE_WIDTH), 
                IL.ilGetInteger(IL.IL_IMAGE_HEIGHT), 0, GL11.GL_RGB, GL11.GL_UNSIGNED_BYTE, scratch);

        return buf.get(0); // Return Image Address In Memory
    }
}

File: c_sharp/lesson20/lesson20.cs
/* 
 * NeHe Lesson 20 - Masking
 * http://nehe.gamedev.net/data/lessons/lesson.asp?lesson=20
 * 
 * Ported to C# by Brian Holley, February, 2004
 * http://tachyon.unl.edu
 * 
 * Uses the CsGL library from http://csgl.sourceforge.net
 * Some CsGL setup and image code from NeHe Lesson 06 C# port by Sabine Felsinger
 */

using System;
using System.Drawing;
using System.Windows.Forms;
using CsGL.OpenGL;

namespace Lesson20
{
	public class MainForm : System.Windows.Forms.Form	
	{
		public Lesson20.LessonView view;

		public MainForm()
		{
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(640, 480);
			this.Name = "MainForm";
			this.Text = "NeHe Lesson 20 - C#";
			this.view = new Lesson20.LessonView();
			this.view.Parent = this;
			this.view.Dock = DockStyle.Fill;
			this.Show();
		}

		static void Main() 
		{
			MainForm form = new MainForm();

			while ((!form.view.finished) && (!form.IsDisposed))
			{
				form.view.glDraw();
				form.Refresh();
				Application.DoEvents();
			}

			form.Dispose();
		}
	}
	
	public class LessonView : OpenGLControl
	{
		public bool masking = true;				// Masking On/Off
		public bool mp = false;					// M Pressed?
		public bool sp = false;					// Space Pressed?
		public bool scene = false;				// Which Scene To Draw
		public float roll = 0.0f;				// Rolling Texture
		
		public uint[] texture = new uint[5];	// Storage For Our Five Textures
		
		public bool finished;
        
		public LessonView() : base()
		{
			this.KeyDown += new KeyEventHandler(LessonView_KeyDown);
			this.KeyUp += new KeyEventHandler(LessonView_KeyUp);
			this.finished = false;
		}

		protected override void InitGLContext() 
		{
			LoadTextures();

			GL.glEnable(GL.GL_TEXTURE_2D);									// Enable Texture Mapping
			GL.glShadeModel(GL.GL_SMOOTH);									// Enable Smooth Shading
			GL.glClearColor(0.0f, 0.0f, 0.0f, 0.0f);						// Black Background
			GL.glClearDepth(1.0f);											// Depth Buffer Setup
			GL.glEnable(GL.GL_DEPTH_TEST);									// Enables Depth Testing
		}

		protected bool LoadTextures()
		{
			Bitmap[] image = new Bitmap[5];
			string[] file = {@"Data\logo.bmp", @"Data\mask1.bmp", @"Data\image1.bmp", @"Data\mask2.bmp", @"Data\image2.bmp"};
			try
			{
				// If the files don't exist or can't be found, an ArgumentException is thrown instead of
				// just returning null
				for (int i=0; i < file.Length; i++)
					image[i] = new Bitmap(file[i]);
			} 
			catch (System.ArgumentException)
			{
				MessageBox.Show("Could not load data files.  Please make sure that Data is a subfolder from where the application is running.", "Error", MessageBoxButtons.OK);
				this.finished = true;
			}
			if (!this.finished)
			{
				GL.glGenTextures(image.Length, this.texture);
			
				for (int i=0; i < image.Length; i++)
				{
					image[i].RotateFlip(RotateFlipType.RotateNoneFlipY);
					System.Drawing.Imaging.BitmapData bitmapdata;
					Rectangle rect = new Rectangle(0, 0, image[i].Width, image[i].Height);

					bitmapdata = image[i].LockBits(rect, System.Drawing.Imaging.ImageLockMode.ReadOnly, System.Drawing.Imaging.PixelFormat.Format24bppRgb);

					// Create Linear Filtered Texture
					GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[i]);
					GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MAG_FILTER, GL.GL_LINEAR);
					GL.glTexParameteri(GL.GL_TEXTURE_2D, GL.GL_TEXTURE_MIN_FILTER, GL.GL_LINEAR);
					GL.glTexImage2D(GL.GL_TEXTURE_2D, 0, (int)GL.GL_RGB, image[i].Width, image[i].Height, 0, GL.GL_BGR_EXT, GL.GL_UNSIGNED_BYTE, bitmapdata.Scan0);

					image[i].UnlockBits(bitmapdata);
					image[i].Dispose();
				}
				return true;
			}
			return false;
		}

		public override void glDraw()
		{
			GL.glClear(GL.GL_COLOR_BUFFER_BIT | GL.GL_DEPTH_BUFFER_BIT);
			GL.glLoadIdentity();
			
			GL.glTranslatef(0.0f, 0.0f, -2.0f);						// Move Into The Screen 5 Units

			GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[0]);			// Select Our Logo Texture
			GL.glBegin(GL.GL_QUADS);									// Start Drawing A Textured Quad
			GL.glTexCoord2f(0.0f, -this.roll+0.0f); GL.glVertex3f(-1.1f, -1.1f,  0.0f);	// Bottom Left
			GL.glTexCoord2f(3.0f, -this.roll+0.0f); GL.glVertex3f( 1.1f, -1.1f,  0.0f);	// Bottom Right
			GL.glTexCoord2f(3.0f, -this.roll+3.0f); GL.glVertex3f( 1.1f,  1.1f,  0.0f);	// Top Right
			GL.glTexCoord2f(0.0f, -this.roll+3.0f); GL.glVertex3f(-1.1f,  1.1f,  0.0f);	// Top Left
			GL.glEnd();											// Done Drawing The Quad

			GL.glEnable(GL.GL_BLEND);									// Enable Blending
			GL.glDisable(GL.GL_DEPTH_TEST);							// Disable Depth Testing

			if (this.masking)										// Is Masking Enabled?
				GL.glBlendFunc(GL.GL_DST_COLOR, GL.GL_ZERO);				// Blend Screen Color With Zero (Black)
			
			if (this.scene)											// Are We Drawing The Second Scene?
			{
				GL.glTranslatef(0.0f, 0.0f, -1.0f);					// Translate Into The Screen One Unit
				GL.glRotatef(this.roll*360, 0.0f, 0.0f, 1.0f);				// Rotate On The Z Axis 360 Degrees.
				if (this.masking)									// Is Masking On?
				{
					GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[3]);	// Select The Second Mask Texture
					GL.glBegin(GL.GL_QUADS);							// Start Drawing A Textured Quad
					GL.glTexCoord2f(0.0f, 0.0f); GL.glVertex3f(-1.1f, -1.1f,  0.0f);	// Bottom Left
					GL.glTexCoord2f(1.0f, 0.0f); GL.glVertex3f( 1.1f, -1.1f,  0.0f);	// Bottom Right
					GL.glTexCoord2f(1.0f, 1.0f); GL.glVertex3f( 1.1f,  1.1f,  0.0f);	// Top Right
					GL.glTexCoord2f(0.0f, 1.0f); GL.glVertex3f(-1.1f,  1.1f,  0.0f);	// Top Left
					GL.glEnd();									// Done Drawing The Quad
				}

				GL.glBlendFunc(GL.GL_ONE, GL.GL_ONE);					// Copy Image 2 Color To The Screen
				GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[4]);		// Select The Second Image Texture
				GL.glBegin(GL.GL_QUADS);								// Start Drawing A Textured Quad
				GL.glTexCoord2f(0.0f, 0.0f); GL.glVertex3f(-1.1f, -1.1f,  0.0f);	// Bottom Left
				GL.glTexCoord2f(1.0f, 0.0f); GL.glVertex3f( 1.1f, -1.1f,  0.0f);	// Bottom Right
				GL.glTexCoord2f(1.0f, 1.0f); GL.glVertex3f( 1.1f,  1.1f,  0.0f);	// Top Right
				GL.glTexCoord2f(0.0f, 1.0f); GL.glVertex3f(-1.1f,  1.1f,  0.0f);	// Top Left
				GL.glEnd();										// Done Drawing The Quad
			}
			else												// Otherwise
			{
				if (this.masking)									// Is Masking On?
				{
					GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[1]);	// Select The First Mask Texture
					GL.glBegin(GL.GL_QUADS);							// Start Drawing A Textured Quad
					GL.glTexCoord2f(this.roll+0.0f, 0.0f); GL.glVertex3f(-1.1f, -1.1f,  0.0f);	// Bottom Left
					GL.glTexCoord2f(this.roll+4.0f, 0.0f); GL.glVertex3f( 1.1f, -1.1f,  0.0f);	// Bottom Right
					GL.glTexCoord2f(this.roll+4.0f, 4.0f); GL.glVertex3f( 1.1f,  1.1f,  0.0f);	// Top Right
					GL.glTexCoord2f(this.roll+0.0f, 4.0f); GL.glVertex3f(-1.1f,  1.1f,  0.0f);	// Top Left
					GL.glEnd();									// Done Drawing The Quad
				}

				GL.glBlendFunc(GL.GL_ONE, GL.GL_ONE);					// Copy Image 1 Color To The Screen
				GL.glBindTexture(GL.GL_TEXTURE_2D, this.texture[2]);		// Select The First Image Texture
				GL.glBegin(GL.GL_QUADS);								// Start Drawing A Textured Quad
				GL.glTexCoord2f(roll+0.0f, 0.0f); GL.glVertex3f(-1.1f, -1.1f,  0.0f);	// Bottom Left
				GL.glTexCoord2f(roll+4.0f, 0.0f); GL.glVertex3f( 1.1f, -1.1f,  0.0f);	// Bottom Right
				GL.glTexCoord2f(roll+4.0f, 4.0f); GL.glVertex3f( 1.1f,  1.1f,  0.0f);	// Top Right
				GL.glTexCoord2f(roll+0.0f, 4.0f); GL.glVertex3f(-1.1f,  1.1f,  0.0f);	// Top Left
				GL.glEnd();										// Done Drawing The Quad
			}

			GL.glEnable(GL.GL_DEPTH_TEST);							// Enable Depth Testing
			GL.glDisable(GL.GL_BLEND);								// Disable Blending

			this.roll += 0.002f;										// Increase Our Texture Roll Variable
			if (this.roll > 1.0f)										// Is Roll Greater Than One
				this.roll -= 1.0f;										// Subtract 1 From Roll
		}

		protected override void OnSizeChanged(EventArgs e)
		{
			base.OnSizeChanged(e);
			Size s = Size;

			if (s.Height == 0)
				s.Height = 1;

			GL.glViewport(0, 0, s.Width, s.Height);

			GL.glMatrixMode(GL.GL_PROJECTION);
			GL.glLoadIdentity();
			GL.gluPerspective(45.0f, (double)s.Width /(double) s.Height, 0.1f, 100.0f);	
			GL.glMatrixMode(GL.GL_MODELVIEW);
			GL.glLoadIdentity();
		}

		protected void LessonView_KeyDown(object Sender, KeyEventArgs e)
		{
			if (e.KeyCode == Keys.Escape)					// Finish the application if the escape key was pressed
				this.finished = true;
			else if (e.KeyCode == Keys.Space && !this.sp)	// Is Space Being Pressed?
			{
				this.sp = true;								// Tell Program Spacebar Is Being Held
				this.scene = !this.scene;					// Toggle From One Scene To The Other
			}
			else if (e.KeyCode == Keys.M && !this.mp)		// Is M Being Pressed?
			{
				this.mp = true;								// Tell Program M Is Being Held
				this.masking = !this.masking;				// Toggle Masking Mode OFF/ON
			}
		}

		private void LessonView_KeyUp(object sender, KeyEventArgs e)
		{
			if (e.KeyCode == Keys.Space)
				this.sp = false;
			else if (e.KeyCode == Keys.M)
				this.mp = false;
		}
	}
}